### PR TITLE
check block hash when import

### PR DESCRIPTION
**Pull request by KickSeason**
avoid continuing sync when fork.

### CODE REVIEW

**Comment by vncoelho** [comment]
Hi @kickseason, I agree with this for NEO2x, right? It would skip import.

However, for NEO3 this may not be needed.
However, i do not know if we should merge this for NEO 2 also, because it does not make much sense when importing.

**Comment by KickSeason** [comment]
https://github.com/neo-project/neo/pull/1047#issuecomment-524138391

File: neo/Ledger/Blockchain.cs [hunk 1]
```diff
@@ -228,7 +228,7 @@ private void OnImport(IEnumerable<Block> blocks)
             foreach (Block block in blocks)
             {
                 if (block.Index <= Height) continue;
-                if (block.Index != Height + 1)
+                if (block.Index != Height + 1 || block.PrevHash != CurrentBlockHash)
```
Comment (erikzhang) [inline neo/Ledger/Blockchain.cs]: We can check the first hash only, instead of check it for every block.

**Comment by shargon** [comment]
How this will protect when a forked block arrive to me?

File: neo/Ledger/Blockchain.cs [hunk 2]
```diff
@@ -228,7 +228,7 @@ private void OnImport(IEnumerable<Block> blocks)
...
                     throw new InvalidOperationException();
```
Comment (lock9) [inline neo/Ledger/Blockchain.cs]: What happens after the exception? Shouldn't we use some message inside this exception? Something like "Invalid block during import" or equivalent?

File: neo/Ledger/Blockchain.cs (hunk 1 above)
Comment (vncoelho) [reply neo/Ledger/Blockchain.cs]: Exactly, I was going to propose this as well, @erikzhang, connected minds.

File: neo/Ledger/Blockchain.cs (hunk 2 above)
Comment (KickSeason) [reply neo/Ledger/Blockchain.cs]: Will do.

File: neo/Ledger/Blockchain.cs (hunk 1 above)
Comment (KickSeason) [reply neo/Ledger/Blockchain.cs]: What if  there is wrong block in middle of chain data package? Should we move this check logic in ImporBlocksPlugin?

**Comment by KickSeason** [comment]
@shargon  This won't stop fork block via p2p. But can protect when syncing fork block from offline chain data package. Once a node forked it can't continue syncing block.But now it can do via import blocks now. Beacuse import block only check index.

**Comment by shargon** [comment]
We should import using "verify"

**Comment by shargon** [comment]
We should import using "verify"

**Comment by KickSeason** [comment]
@shargon I agree. Will "verify" slow chain data syncing speed a lot?  If we all  agree I can add the import "verify".

**Comment by shargon** [comment]
Slow, but secure

**Comment by lock9** [comment]
Does anybody has a broken chain package?

**Comment by vncoelho** [comment]
I recomender when that was removed.

I am not sure, that should be optional.

**Comment by KickSeason** [comment]
We can add  this parameter in ImportBlocks plugin. Users can choose to check hash or the whole block.

File: neo/Ledger/Blockchain.cs [hunk 3]
```diff
@@ -228,7 +228,7 @@ private void OnImport(IEnumerable<Block> blocks)
             foreach (Block block in blocks)
```
Comment (shargon) [inline neo/Ledger/Blockchain.cs]: Order by Index?

File: neo/Ledger/Blockchain.cs (hunk 1 above)
Comment (shargon) [reply neo/Ledger/Blockchain.cs]: I think that we should verify every blocks, we need to ensure that the signature is the expected one, slow, but secure.

**Comment by KickSeason** [comment]
I think we leave neo2 and add this feature on neo3.

**Pull request closed without merging**